scrapy/item.py

```diff
@@ -28,7 +28,12 @@ class BaseItem(object_ref):
     Unlike instances of :class:`dict`, instances of :class:`BaseItem` may be
     :ref:`tracked <topics-leaks-trackrefs>` to debug memory leaks.
     """
-    pass
```

Comment:
It would be nice to update BaseItem docstring. 

Do you know if there are other places in docs which need to be updated, after BaseItem deprecation? I guess it would require searching for BaseItem in Scrapy codebase and documentation.

One more question: any idea on how we would be able to remove it completely? :) There is code in various middlewares which check for `isinstance(item, BaseItem)`; would we be able to issue a deprecation warning for such checks in future?

Reply:
Regarding the latter, [it looks like it may be doable](https://stackoverflow.com/questions/9008444/how-to-warn-about-class-name-deprecation/9008509#comment91699108_26130976) (TIL).

Reply:
I updated the docstring and updated a few `BaseItem` occurrences I missed before. This will probably create some conflicts with #3881, mostly regarding docs, but that's fine.

I couldn't find any occurrences of `BaseItem` in middlewares, just in other places which should work just fine by replacing `BaseItem` with `Item`:
```
$ grep -rn  "BaseItem\b" scrapy
scrapy/contracts/default.py:3:from scrapy.item import BaseItem
scrapy/contracts/default.py:54:        'item': (BaseItem, dict),
scrapy/contracts/default.py:55:        'items': (BaseItem, dict),
scrapy/contracts/default.py:106:            if isinstance(x, (BaseItem, dict)):
scrapy/utils/serialize.py:8:from scrapy.item import BaseItem
scrapy/utils/serialize.py:29:        elif isinstance(o, BaseItem):
scrapy/utils/misc.py:17:from scrapy.item import BaseItem
scrapy/utils/misc.py:20:_ITERABLE_SINGLE_VALUES = dict, BaseItem, str, bytes
scrapy/commands/parse.py:8:from scrapy.item import BaseItem
scrapy/commands/parse.py:120:            if isinstance(x, (BaseItem, dict)):
scrapy/core/scraper.py:17:from scrapy.item import BaseItem
scrapy/core/scraper.py:194:        elif isinstance(output, (BaseItem, dict)):
scrapy/core/scraper.py:203:            logger.error('Spider must return Request, BaseItem, dict or None, '
scrapy/shell.py:16:from scrapy.item import BaseItem
scrapy/shell.py:29:    relevant_classes = (Crawler, Spider, Request, Response, BaseItem,
scrapy/item.py:17:class BaseItem(object_ref):
scrapy/item.py:23:        if issubclass(cls, BaseItem) and not (issubclass(cls, Item) or issubclass(cls, DictItem)):
scrapy/item.py:24:            warn('scrapy.item.BaseItem is deprecated, please use scrapy.item.Item instead',
scrapy/item.py:26:        return super(BaseItem, cls).__new__(cls, *args, **kwargs)
scrapy/item.py:60:class DictItem(MutableMapping, BaseItem):
scrapy/item.py:105:    __hash__ = BaseItem.__hash__
scrapy/exporters.py:15:from scrapy.item import BaseItem
scrapy/exporters.py:315:        if isinstance(value, BaseItem):
```

Sorry, but TBH I'm not sure I fully understand the last question. Could you guys elaborate?

Reply:
This should be addressed by 5256eae60d3685de51c1f3891abe157e15d14def. Introduced yet another (private) class, `_BaseItem`, as a way to avoid deprecation warnings when using `isinstance` internally. Most of the imports accross the codebase should go away after #3881, replaced by `itemadapter.is_item`.